### PR TITLE
Improved metrics

### PR DESCRIPTION
**Pull request by Adaickalavan**
- Made the metrics module configurable by supplying parameters through a `Params` class.
- `Params` allows to specify vehicles to be ignored in `dist_to_obstacles` cost function. This would be applicable in platooning tasks.
- Unified computation of `dist_to_destination` (previously known as `completion`) and `steps` (i.e., time taken) as functions inside the cost functions module, instead of computing them separately in a different module.
- In the metrics module, the records which is the raw metrics data and the scoring which is the formula to compute the final results are now separated to provided greater flexibility for applying metrics to different environments.
- Changed `benchmark_runner_v0.py` to only average records across scenarios within an environment. Records are not averaged across enviroments, because the scoring formula may differ in different environments.
- Fixed CI tests for metrics.

### CODE REVIEW

File: smarts/env/gymnasium/wrappers/metric/params.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    """Parameters for wrong way cost function."""
+
+    active: bool = True
+    """If True, enables computation of coresponding cost function. Else, 
+    disabled.
+    """
+
+
+@dataclass
+class Params:
```
[the remark this answers is not in the document]
Comment (Adaickalavan) [reply smarts/env/gymnasium/wrappers/metric/params.py]: Yes, changed all dataclasses to frozen.

File: smarts/env/gymnasium/wrappers/metric/params.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+@dataclass
+class DistToObstacles:
+    """Parameters for distance to obstacles cost function."""
+
+    active: bool = True
+    """If True, enables computation of coresponding cost function. Else, 
+    disabled.
+    """
+    ignore: List[str] = field(default_factory=lambda: [])
```
[the remark this answers is not in the document]
Comment (Adaickalavan) [reply smarts/env/gymnasium/wrappers/metric/params.py]: Changed as suggested.

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,12 +21,18 @@ Copy and pasting the git commit messages is __NOT__ enough.
 - The trap manager, `TrapManager`, is now a subclass of `ActorCaptureManager`.
 - Considering lane-change time ranges between 3s and 6s, assuming a speed of 13.89m/s, the via sensor lane acquisition range was increased from 40m to 80m, for better driving ability.
 - Modified naming of benchmark used in NeurIPS 2022 from driving-smarts-competition-env to driving-smarts-v2022.
+- Made the metrics module configurable by supplying parameters through a `Params` class.
+- `Params` allows to specify vehicles to be ignored in `dist_to_obstacles` cost function. This would be applicable in platooning tasks.
```
[the remark this answers is not in the document]
Comment (Adaickalavan) [reply CHANGELOG.md]: I have rephrased the changelog entry.

**Review by saulfield — APPROVED** [review]

File: smarts/benchmark/driving_smarts/v2023/config_3.yaml [hunk 4]
```diff
@@ -20,4 +20,4 @@ benchmark:
         - scenarios/sumo/platoon/merge_exit_sumo_t_agents_1
       kwargs:
         seed: 42
-      # metric_formula: smarts/benchmark/driving_smarts/v2023/metric_formula_platoon.py
+      metric_formula: smarts/benchmark/driving_smarts/v2023/metric_formula_platoon.py
```
Comment (Gamenot) [inline smarts/benchmark/driving_smarts/v2023/config_3.yaml:23]: I think we should mention this in the changelog if not already.

**Review by Gamenot — APPROVED** [review]

File: smarts/benchmark/driving_smarts/v2023/config_3.yaml (hunk 4 above)
Comment (Adaickalavan) [reply smarts/benchmark/driving_smarts/v2023/config_3.yaml:23]: Previously changelog stated
```
- In the metrics module, the records which is the raw metrics data and the scoring which is the formula to compute the final results are now separated to provided greater flexibility for applying metrics to different environments.
```
Now, another entry, as follows is added.
```
- Benchmark listing may specify specialised metric formula for each benchmark.
```

**Pull request merged**